src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -623,11 +624,26 @@ private static string Encode5987(StringSegment input)
                         i++;
                     }
 
-                    var unicodePart = input.Value.Substring(startPos, i - startPos + 1);
-                    var bytes = Encoding.UTF8.GetBytes(unicodePart);
-                    foreach (byte b in bytes)
+                    var unicodePart = ((ReadOnlySpan<char>)input).Slice(startPos, i - startPos + 1);
+                    int bytesCount = Encoding.UTF8.GetByteCount(unicodePart);
+
+                    try
                     {
-                        HexEscape(builder, (char)b);
+                        buffer = ArrayPool<byte>.Shared.Rent(bytesCount);
+                        //Buffer from the pool may be larger than we need
+                        var exactBuffer = buffer.AsSpan().Slice(0, bytesCount);
```

Comment:
No need to slice here. `Encoding.UTF8.GetBytes` returns the written bytes, use that value for slicing.
Also above in L628 you can use `GetMaxByteCount` to get a big enough buffer*, which has cheaper cost that `GetByteCount`.

Why not stackalloc? I guess that the buffer-sizes will be rather small, so stackalloc is a perfect fit.

\* in terms of the ArrayPool the actual size will be (implementation detail) rounded up to the next power of two anyway if the rental can be served from the pool

Reply:
To use stackalloc, we need to estimate buffer size upfront which looks like reasonable tradeoff for simplicity and performance we get on the other hand. New update is coming.